### PR TITLE
add /Car endpoint

### PR DESCRIPTION
**Pull request by willscott**
for getting the ipld dag of finalized tasks

### CODE REVIEW

File: controller/state/statedb.go [hunk 1]
```diff
@@ -8,12 +8,14 @@ import (
...
 	"fmt"
 	"io"
+	"io/ioutil"
 	"net/http"
 	"sync"
 	"time"
 
 	"github.com/filecoin-project/dealbot/metrics"
 	"github.com/filecoin-project/dealbot/tasks"
+	blockformat "github.com/ipfs/go-block-format"
```
Comment (hannahhoward) [inline controller/state/statedb.go:18]: non-blocking: for better or worse, in most repos the shortname used here is 'blocks'

File: controller/tasks.go [hunk 2]
```diff
@@ -241,6 +245,40 @@ func (c *Controller) getTaskHandler(w http.ResponseWriter, r *http.Request) {
...
+		log.Errorw("get task DB error", "err", err.Error())
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+
+	ssb := builder.NewSelectorSpecBuilder(basicnode.Prototype__Any{})
+	ss := ssb.ExploreRecursive(selector.RecursionLimitNone(), ssb.ExploreUnion(
+		ssb.Matcher(),
```
Comment (hannahhoward) [inline controller/tasks.go:268]: Non-blocking: FYI, for the purposes of CAR writing, the ExploreUnion and the Matcher are unnecessary

File: controller/tasks.go [hunk 3]
```diff
@@ -241,6 +245,40 @@ func (c *Controller) getTaskHandler(w http.ResponseWriter, r *http.Request) {
...
+		ssb.ExploreAll(ssb.ExploreRecursiveEdge()),
+	))
+	root := car.Dag{
+		Root:     rootCid,
+		Selector: ss.Node(),
+	}
+	sc := car.NewSelectiveCar(r.Context(), store, []car.Dag{root})
```
Comment (hannahhoward) [inline controller/tasks.go:275]: side note: we should make a version of selective car that actually just takes an IPLD loader -- under the hood, this is actually converting the reader BACK to a loader, which is funny case you converted a loader to a reader to get this to work :)

File: controller/controller.go [hunk 4]
```diff
@@ -121,6 +121,7 @@ func NewWithDependencies(listener, graphqlListener net.Listener, recorder metric
 	r.HandleFunc("/status", srv.reportStatusHandler).Methods("POST")
 	r.HandleFunc("/tasks/{uuid}", srv.updateTaskHandler).Methods("PATCH")
 	r.HandleFunc("/tasks/{uuid}", srv.getTaskHandler).Methods("GET")
+	r.HandleFunc("/car", srv.carHandler).Methods("GET")
```
Comment (hannahhoward) [inline controller/controller.go:124]: This strikes me as being worthy of an HTTP test in http_test.go :)

**Review by hannahhoward — APPROVED** [review]
Would be great to add a test of the endpoint in the HTTP test file (vaguely assuming we can populate the database somehow?)

Not gonna block on it but consider it a strong recommend? At least for a seperate ticket?

**Comment by willscott** [comment]
Added at test. Note we're now on a weird offshoot ipld prime commit i made to backport functional `dag-json` to the v0.7 we're stuck at

**Pull request merged**